File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/processor/ValidatableTestStepProcessor.java
package org.skellig.teststep.processing.processor;

import org.skellig.teststep.processing.model.TestStep;
import org.skellig.teststep.processing.model.ValidationDetails;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.validation.TestStepResultValidator;
import org.skellig.teststep.reader.exception.ValidationException;

import java.util.Optional;

public abstract class ValidatableTestStepProcessor<T extends TestStep> implements TestStepProcessor<T> {

    protected static final String RESULT_SAVE_SUFFIX = ".result";

    protected final TestScenarioState testScenarioState;
    protected final TestStepResultValidator validator;

    protected ValidatableTestStepProcessor(TestScenarioState testScenarioState, TestStepResultValidator validator) {
        this.testScenarioState = testScenarioState;
        this.validator = validator;
    }

    protected void validate(TestStep testStep, Object actualResult) {
        testStep.getValidationDetails()
                .ifPresent(validationDetails -> {
                    Optional<String> testStepId = validationDetails.getTestStepId();
                    if (testStepId.isPresent()) {
                        Optional<Object> actualResultFromAnotherStep = getLatestResultOfTestStep(testStepId.get());
                        if (actualResultFromAnotherStep.isPresent()) {
                            validate(testStep.getId(), validationDetails, actualResultFromAnotherStep.get());
                        } else {
                            throw new ValidationException(String.format("Result from test step with id '%s' was not found " +
                                    "in Test Scenario State", testStepId.get()));
                        }
                    } else {
                        validate(testStep.getId(), validationDetails, actualResult);
                    }
                });
    }

    private void validate(String testStepId, ValidationDetails validationDetails, Object actualResultFromAnotherStep) {
        try {
            validator.validate(validationDetails.getExpectedResult(), actualResultFromAnotherStep);
        } catch (ValidationException ex) {
            throw new ValidationException(ex.getMessage(), testStepId);
        }
    }

    protected Optional<Object> getLatestResultOfTestStep(String testStepId) {
        return testScenarioState.get(testStepId + RESULT_SAVE_SUFFIX);
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/processor/ValidatableTestStepProcessor.kt
package org.skellig.teststep.processing.processor

import org.skellig.teststep.processing.exception.ValidationException
import org.skellig.teststep.processing.model.DefaultTestStep
import org.skellig.teststep.processing.model.ValidationNode
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processing.util.error
import org.skellig.teststep.processing.util.info
import org.skellig.teststep.processing.util.logger

/**
 * This is a basic processor for tests steps whose result can be validated.
 */
abstract class ValidatableTestStepProcessor<T : DefaultTestStep>(protected val testScenarioState: TestScenarioState) : TestStepProcessor<T> {

    private val log = logger<ValidatableTestStepProcessor<*>>()

    /**
     * Validate test step actual result with expected one.
     * The expected result is part of `validation` section of the test step.
     *
     * If no validation details provided then it does nothing.
     *
     * If validation fails, then it throws `ValidationException`
     * @see ValidationException
     */
    @Throws(ValidationException::class)
    protected open fun validate(testStep: T, actualResult: Any?) {
        testStep.validationDetails?.let { expectedResult ->
            validate(testStep, expectedResult, actualResult)
        }
    }

    /**
     * Checks whether the result of processing of the test step matches with the expected
     * result, provided in its validation details.
     */
    protected open fun isValid(testStep: T, actualResult: Any?): Boolean =
        try {
            if (testStep.validationDetails != null) {
                validate(testStep, testStep.validationDetails, actualResult)
                true
            } else actualResult != null
        } catch (ex: Exception) {
            log.error(testStep, "Validation failed: ${ex.message}")
            false
        }

    protected open fun validate(testStep: T, expectedResult: ValidationNode, actualResult: Any?) {
        try {
            log.info(testStep, "Start to validate the result of processed test step '${testStep.name}'")
            expectedResult.validate(actualResult)
        } catch (ex: ValidationException) {
            throw ValidationException(ex.message)
        }
    }

}